### DISCUSSION TITLE
v0.2.7 域名嗅探不支持http的请求

### DISCUSSION THREAD

**Discussion by liuxc0116** [eycorsican/leaf#94] [Q&A]
linux 下 v0.2.7设置域名过滤规则
```
DOMAIN-SUFFIX, baidu.com, Proxy
```

使用curl请求
```
curl https://www.baidu.com/
```
https可以通过代理

```
curl http://www.baidu.com/
```
http不通过代理

不知道域名嗅探的原理是什么，为什么https可以域名过滤，http反而不行

**Comment by eycorsican** [comment]
HTTP 协议嗅探没做

**Reply by liuxc0116** [reply]
能简单说下，tun数据处理的流程和域名嗅探的过程吗？rust代码读起来有点困难。

**Reply by eycorsican** [reply]
以 TCP 上行链路为例，从[这里](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/proxy/tun/inbound.rs#L114)读 TUN 的 IP 包发给 Netstack，经过 TCP 握手（就是来回读写 3 个 IP 包）后可以 [accept](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/proxy/tun/netstack/tcp_listener_impl.rs#L19) 到一个 TcpStream，这个 TcpStream [在这里接收](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/proxy/tun/netstack/stack_impl.rs#L102)后，会 [dispatch](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/proxy/tun/netstack/stack_impl.rs#L127) 给相应的 outbound。

dispatch stream 时，在把 stream 给到 outbound 前，会有机会查看/修改 stream 里的内容，嗅探功能就是在这个点上[插入](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/app/dispatcher.rs#L211)，除开相应协议（TLS/HTTP）细节不说，嗅探功能就是一个简单的实现 AsyncRead +  AsyncWrite 的 stream，主要是在 [AsyncRead](https://github.com/eycorsican/leaf/blob/01fb4d64fa41d0402f3a5ae3d8539e754b91a7a4/leaf/src/common/sniff.rs#L158) 中先查看数据内容再写数据。
